zio-profiling/jvm/src/main/scala/zio/profiling/causal/SamplingState.scala

```diff
@@ -0,0 +1,28 @@
+package zio.profiling.causal
+
+sealed private trait SamplingState
+
+private object SamplingState {
+  final case class Warmup(
+    until: Long
+  ) extends SamplingState
+
+  final case class ExperimentPending(
+    iteration: Int,
+    results: List[ExperimentResult]
+  ) extends SamplingState
+
+  final case class ExperimentInProgress(
+    experiment: Experiment,
+    iteration: Int,
```

Comment:
In principle, we could make iteration mutable to safe allocations, but this the whole sampling loop is not really in the hot path